docs/RFCS/table_descriptor_lease.md

```diff
@@ -0,0 +1,309 @@
+the node has a lease for the descriptor.
+
+When performing a schema change operation, the modification to the
+table descriptor is performed using the `Publish` operation which will
+fail if the version is invalid (i.e. too old) or if there are already
+two active versions of the descriptor in the cluster. On success, the
+`Publish` operation will write the new descriptor to the KV store and
+notify all watches of the new descriptor which will cause those nodes
+to stop using the old descriptor for new operations. As existing
+operations complete they will release their local reference to the old
+descriptor which will cause the reference count to drop to 0 and
+result in the lease being released. When all of the leases for the old
+descriptor have been released, the lease master will send out a
+notification to the watches that the version is `DEAD`.
+
+Leases will be granted for a duration measured in minutes (we'll
+assume 5m for the rest of this doc, though experimentation may tune
+this number). A node will renew recently used descriptors before their
+leases expires and ensure that when handing out a descriptor for local
+usage the lease will not expire for at least `<leaseDuration>/2`
+minutes. When a node starts it will issue a `Release` message with the
+`all` field set to true to indicate it wants to release any leases
+held by its previous incarnation. This will ensure that a transient
+failure does not require the lease expiration to occur before allowing
+a schema modification. When a node dies permanently or becomes
+unresponsive (e.g. detached from the network) the lease master will
+have to wait for any leases that node held to expire. This will result
+in an inability to perform more than one schema modification step to
+the descriptors referred to by those leases. The maximum amount of
+time this condition can exist is the lease duration.
+
+If the lease master dies a new master will be "elected" within
+20s. Lease state will be stored in a `leaseTable` that will be read
+when the master starts:
+
+```sql
+CREATE TABLE leaseTable (
+  DescID     INT,
+  Version    INT,
+  NodeID     INT,
+  Expiration TIMESTAMP,
+  PRIMARY KEY (DescID, Version, NodeID)
+)
+```
+
+Entries in the lease table will be added and removed as leases are
+acquired and released. A background goroutine will periodically delete
+expired leases. The master will maintain the invariant that for a
+given descriptor there are only two active versions (i.e. non-expired
+versions) in the lease table and that the two versions correspond to
+the table descriptor in the KV store and its immediate predecessor.
+
+When the lease master is unavailable (because it has died and a new
+master has not yet started), new leases cannot be granted but existing
+leases remain valid until they expire.
+
+# Drawbacks
+
+* This is complex. Perhaps overly complex. Where can it be simplified?
+
+* The watch mechanism is described with the assumption of streaming
```

Comment:
Yes, gossip does have a delay (production default is currently 2s/hop, overridden to 10ms/hop in tests). This may be acceptable since it gets the delay for a small cluster down to something reasonable, although it doesn't help small tables in a large multi-tenant cluster.

I think we could also add a high-priority flag to gossip that would let us bypass this delay completely. I'd prefer to use/enhance gossip if we can instead of introducing a new mechanism.

The `DEAD` notification wouldn't fit cleanly into gossip, though. Since that's only of interest to the node attempting the schema change, though, we could reduce the watch RPC to this case instead of using it for all lease holders. Or the node doing the schema change could just poll to see when it can move on to the next phase.